icu4c/source/i18n/unitconverter.cpp

```diff
@@ -15,6 +18,164 @@
+        }
+
+        bool shouldFlip = power < 0; // This means that after applying the absolute power, we should flip
+                                     // the Numerator and Denominator.
+
+        factorNum = std::pow(factorNum, std::abs(power));
+        factorDen = std::pow(factorDen, std::abs(power));
+
+        if (shouldFlip) {
+            // Flip Numerator and Denominator.
+            std::swap(factorNum, factorDen);
+        }
+    }
+
+    // Flip the `Factor`, for example, factor= 2/3, flippedFactor = 3/2
+    void flip() {
+        std::swap(factorNum, factorDen);
+
+        for (int i = 0; i < CONSTANTS_COUNT; i++) {
+            constants[i] *= -1;
+        }
+    }
+
+    // Apply SI prefix to the `Factor`
+    void applySiPrefix(UMeasureSIPrefix siPrefix) {
+        if (siPrefix == UMeasureSIPrefix::UMEASURE_SI_PREFIX_ONE) return; // No need to do anything
+
+        double siApplied = std::pow(10.0, std::abs(siPrefix));
+
+        if (siPrefix < 0) {
+            factorDen *= siApplied;
+            return;
+        }
+
+        factorNum *= siApplied;
+    }
+
+    void substituteConstants() {
+        double constantsValues[CONSTANTS_COUNT];
+
+        // TODO: Load those constant values from units data.
+        constantsValues[CONSTANT_FT2M] = 0.3048;
+        constantsValues[CONSTANT_PI] = 411557987.0 / 131002976.0;
+        constantsValues[CONSTANT_GRAVITY] = 9.80665;
+        constantsValues[CONSTANT_G] = 6.67408E-11;
+        constantsValues[CONSTANT_LB2KG] = 0.45359237;
+        constantsValues[CONSTANT_GAL_IMP2M3] = 0.00454609;
+
+        for (int i = 0; i < CONSTANTS_COUNT; i++) {
+            if (this->constants[i] == 0) { continue;}
+
+            auto absPower = std::abs(this->constants[i]);
+            SigNum powerSig = this->constants[i] < 0 ? SigNum::NEGATIVE : SigNum::POSITIVE;
+            double absConstantValue = std::pow(constantsValues[i], absPower);
+
+            if (powerSig ==  SigNum::NEGATIVE) { this->factorDen *= absConstantValue;} 
+            else { this->factorNum *= absConstantValue;}
+
+            this->constants[i] = 0;
+        }
+    }
+};
+
+/* Helpers */
+
+using icu::double_conversion::StringToDoubleConverter;
+
+// TODO: Make this a shared-utility function.
+// Returns `double` from a scientific number(i.e. "1", "2.01" or "3.09E+4")
+double strToDouble(StringPiece strNum, UErrorCode &status) {
+    // We are processing well-formed input, so we don't need any special options to
+    // StringToDoubleConverter.
+    StringToDoubleConverter converter(0, 0, 0, "", "");
+    int32_t count;
+    double result = converter.StringToDouble(strNum.data(), strNum.length(), &count);
+    if (count != strNum.length()) { status = U_INVALID_FORMAT_ERROR; }
+
+    return result;
+}
+
+// Returns `double` from a scientific number that could has a division sign (i.e. "1", "2.01", "3.09E+4"
+// or "2E+2/3")
+double strHasDivideSignToDouble(StringPiece strWithDivide, UErrorCode &status) {
```

Comment:
I suggest renaming this function and improving the comment. Would you prefer I make suggestions here, or should I simply go ahead and edit this in a commit or PR of my own?

Considering this: "2E+2/3" - it's not only a scientific number. Maybe "fractionStringToDouble"?